src/Swift.Runtime/src/Swift/Set.cs

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.InteropServices.Swift;
+using Swift.Runtime;
+using Swift.Runtime.InteropServices;
+
+namespace Swift;
+
+/// <summary>
+/// Represents a Swift set payload.
+///
+/// The following diagram illustrates the hierarchy of a Swift set type.
+/// The actual implementation may differ:
+///
+///    struct Set
+///    +-----------------------------------------------------------------------+
+///    |   struct Variant                                                     |
+///    |   +--------------------------------------------------------------+   |
+///    |   |   struct Object                                              |   |
+///    |   |   +------------------------------------------------------+   |   |
+///    |   |   | var rawValue: IntPtr                                 |   |   |
+///    |   |   +------------------------------------------------------+   |   |
+///    |   +--------------------------------------------------------------+   |
+///    +-----------------------------------------------------------------------+
+/// </summary>
+[StructLayout(LayoutKind.Sequential, Size = 8)]
+public struct Variant
+{
+    public IntPtr rawValue;
+}
+
+/// <summary>
+/// Represents a Swift set.
+/// </summary>
+/// <typeparam name="Element">The element type contained in the set.</typeparam>
+public class SwiftSet<Element> : IDisposable, ISwiftObject
+{
+    static nuint _payloadSize = SwiftObjectHelper<SwiftSet<Element>>.GetTypeMetadata().Size;
+    static nuint _elementSize = ElementTypeMetadata.Size;
+
+    // Swift set is a value type and doesn't contain an IntPtr payload
+    private Variant variant;
+    public unsafe void Dispose()
+    {
+        if (variant.rawValue != IntPtr.Zero)
+        {
+            Arc.Release(*(IntPtr*)variant.rawValue);
+            variant.rawValue = IntPtr.Zero;
+            GC.SuppressFinalize(this);
+        }
+    }
+
+    unsafe ~SwiftSet()
+    {
+        Arc.Release(*(IntPtr*)variant.rawValue);
+        variant.rawValue = IntPtr.Zero;
+    }
+    public static nuint PayloadSize => _payloadSize;
+    public Variant Payload => variant;
+    public static nuint ElementSize => _elementSize;
+    public static ProtocolWitnessTable IHashableWitnessTable
```

Comment:
I think that if the only thing we need is an empty set we might not actually need to use `IHashableWitnessTable` at all. We should be good with just passing a nulltpr. We will need to have a valid protocol witness table when we use functions which actually try to hash something.

Reply:
Yes, sounds good.